### PR TITLE
azurestore: Migrate to current Azure SDK

### PR DESCRIPTION
**Pull request by xelat09**
Hi 

Here an effort to migrate to the new azure sdk, manual tests with azurite did work. I could also do some testing with the real thing, once merged.
Maybe you could have a review, why some tests are not green yet and if this is mergable at all :). 

Regards

### CODE REVIEW

**Comment by xelat09** [comment]
I just retried the tests in bash and it worked as expected

File: pkg/azurestore/azureservice.go [hunk 1]
```diff
@@ -67,87 +72,92 @@ type AzBlob interface {
...
 	case "container":
-		containerAccessType = azblob.PublicAccessContainer
+		containerAccessType = container.PublicAccessTypeContainer
 	case "blob":
-		containerAccessType = azblob.PublicAccessBlob
-	case "":
+		containerAccessType = container.PublicAccessTypeBlob
 	default:
-		containerAccessType = azblob.PublicAccessNone
+		containerAccessType = ""
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:114]: The previous default was `azblob.PublicAccessNone`. Was this change intentional?

File: pkg/azurestore/azureservice.go [hunk 2]
```diff
@@ -67,87 +72,92 @@ type AzBlob interface {
 }
 
 type BlockBlob struct {
-	Blob       *azblob.BlockBlobURL
-	AccessTier azblob.AccessTierType
-	Indexes    []int
+	Blob           *blockblob.Client
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:75]: I'm a bit surprised by the field being named `Blob` but the type indicating it's a client object. Can you shed some light on this?

File: pkg/azurestore/azureservice.go [hunk 3]
```diff
@@ -67,87 +72,92 @@ type AzBlob interface {
...
+	cred, err := azblob.NewSharedKeyCredential(config.AccountName, config.AccountKey)
 	if err != nil {
 		return nil, err
 	}
 
-	// Might be limited by the storage account
-	// "" or default inherits the access type from the Storage Account
-	var containerAccessType azblob.PublicAccessType
+	serviceURL := fmt.Sprintf("%s/%s", config.Endpoint, config.ContainerName)
+	retryOpts := policy.RetryOptions{
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:93]: Are those retry options new?

File: pkg/azurestore/azureservice.go [hunk 4]
```diff
@@ -67,87 +72,92 @@ type AzBlob interface {
...
 	case "cool":
-		blobAccessTierType = azblob.AccessTierCool
+		blobAccessTier = to.Ptr(blob.AccessTierCool)
 	case "hot":
-		blobAccessTierType = azblob.AccessTierHot
+		blobAccessTier = to.Ptr(blob.AccessTierHot)
 	case "":
 	default:
-		blobAccessTierType = azblob.DefaultAccessTier
+		blobAccessTier = nil
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:135]: Similar to above, the previous default was `azblob.DefaultAccessTier`. Can we keep that?

File: pkg/azurestore/azureservice.go [hunk 5]
```diff
@@ -67,87 +72,92 @@ type AzBlob interface {
...
+		Blob:           blobClient,
+		Indexes:        []int{},
+		BlobAccessTier: service.BlobAccessTier,
+	}, nil
 }
 
 // Delete the blockBlob from Azure Blob Storage
 func (blockBlob *BlockBlob) Delete(ctx context.Context) error {
-	_, err := blockBlob.Blob.Delete(ctx, azblob.DeleteSnapshotsOptionInclude, azblob.BlobAccessConditions{})
+	_, err := blockBlob.Blob.Delete(ctx, nil)
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:160]: Should we keep `DeleteSnapshotsOptionInclude` or whatever its newer form is?

File: pkg/azurestore/azureservice.go [hunk 6]
```diff
@@ -161,31 +171,22 @@ func (blockBlob *BlockBlob) Upload(ctx context.Context, body io.ReadSeeker) erro
...
-	downloadResponse, err := blockBlob.Blob.Download(ctx, 0, azblob.CountToEnd, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
-
-	// If the file does not exist, it will not return an error, but a 404 status and body
-	if downloadResponse != nil && downloadResponse.StatusCode() == 404 {
-		return nil, handler.ErrNotFound
-	}
+	resp, err := blockBlob.Blob.DownloadStream(ctx, nil)
 	if err != nil {
-		// This might occur when the blob is being uploaded, but a block list has not been committed yet
-		if isAzureError(err, "BlobNotFound") {
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:182]: Should we check for this error as well?

File: pkg/azurestore/azureservice.go [hunk 7]
```diff
@@ -295,14 +290,17 @@ func blockIDIntToBase64(blockID int) string {
 	return blockIDBinaryToBase64(binaryBlockID)
 }
 
-func blockIDBase64ToInt(blockID string) int {
+func blockIDBase64ToInt(blockID *string) int {
 	blockIDBase64ToBinary(blockID)
 	return int(binary.LittleEndian.Uint32(blockIDBase64ToBinary(blockID)))
 }
 
-func isAzureError(err error, code string) bool {
```
Comment (Acconut) [inline pkg/azurestore/azureservice.go:303]: Could we keep this function and inspect errors from Azure by looking into their service code instead of just checking their error message? Using the service code appears more robust in the long term to me.

**Review by Acconut — COMMENTED** [review]
Thank you very much for this! Overall, the PR is looking great and I'm positive we can merge this soon. I'm relieved that no fundamental changes are necessary to the azurestore itself and that most changes are just simple name changes.

File: pkg/azurestore/azureservice.go (hunk 1 above)
Comment (xelat09) [reply pkg/azurestore/azureservice.go:116]: there is no such type anymore, it is just the default behaviour if createOptions is nil. I did refactor according to this

File: pkg/azurestore/azureservice.go (hunk 2 above)
Comment (xelat09) [reply pkg/azurestore/azureservice.go]: yeah, that's ugly.. I tried to hard to not change too much :)
refactored this too

File: pkg/azurestore/azureservice.go (hunk 4 above)
Comment (xelat09) [reply pkg/azurestore/azureservice.go:143]: similar to above, the default behaviour if options is nil, refactored

File: pkg/azurestore/azureservice.go (hunk 5 above)
Comment (xelat09) [reply pkg/azurestore/azureservice.go]: yes, reintroduced that

File: pkg/azurestore/azureservice.go (hunk 7 above)
Comment (xelat09) [reply pkg/azurestore/azureservice.go:303]: yes, sure.. i did that

**Comment by xelat09** [comment]
thank you for the review, I'm still learning, so don't hesitate to point out anything that needs reworking

File: pkg/azurestore/azureservice.go (hunk 6 above)
Comment (xelat09) [reply pkg/azurestore/azureservice.go:182]: yeah fixed that too
